File: archive/cdktf-java/Pr3239/lib/src/main/java/app/constructs/IamConstruct.java
package app.constructs;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRole;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRoleConfig;
import com.hashicorp.cdktf.providers.aws.iam_role_policy.IamRolePolicy;
import com.hashicorp.cdktf.providers.aws.iam_role_policy.IamRolePolicyConfig;
import software.constructs.Construct;

public class IamConstruct extends BaseConstruct {
    private final IamRole lambdaExecutionRole;
    private final ObjectMapper mapper = new ObjectMapper();

    public IamConstruct(final Construct scope, final String id, final String dynamoTableArn, final String s3BucketArn) {
        super(scope, id);

        // Create Lambda execution role
        this.lambdaExecutionRole = new IamRole(this, "lambda-execution-role", IamRoleConfig.builder()
                .name(resourceName("LambdaExecutionRole"))
                .assumeRolePolicy(createAssumeRolePolicy())
                .tags(getTagsWithName("LambdaExecutionRole"))
                .build());

        // Attach CloudWatch Logs policy
        new IamRolePolicy(this, "lambda-logs-policy", IamRolePolicyConfig.builder()
                .name("LambdaLogsPolicy")
                .role(lambdaExecutionRole.getId())
                .policy(createLogsPolicy())
                .build());

        // Attach DynamoDB policy
        new IamRolePolicy(this, "lambda-dynamodb-policy", IamRolePolicyConfig.builder()
                .name(resourceName("LambdaDynamoDBPolicy"))
                .role(lambdaExecutionRole.getId())
                .policy(createDynamoDBPolicy(dynamoTableArn))
                .build());

        // Attach S3 read policy for deployment packages
        new IamRolePolicy(this, "lambda-s3-policy", IamRolePolicyConfig.builder()
                .name(resourceName("LambdaS3Policy"))
                .role(lambdaExecutionRole.getId())
                .policy(createS3Policy(s3BucketArn))
                .build());
    }

    private String createAssumeRolePolicy() {

        ObjectNode policy = mapper.createObjectNode();
        policy.put("Version", "2012-10-17");

        ArrayNode statements = mapper.createArrayNode();
        ObjectNode statement = mapper.createObjectNode();
        statement.put("Effect", "Allow");
        statement.put("Action", "sts:AssumeRole");

        ObjectNode principal = mapper.createObjectNode();
        principal.put("Service", "lambda.amazonaws.com");
        statement.set("Principal", principal);

        statements.add(statement);
        policy.set("Statement", statements);

        return policy.toString();
    }

    private String createLogsPolicy() {
        ObjectNode policy = mapper.createObjectNode();
        policy.put("Version", "2012-10-17");

        ArrayNode statements = mapper.createArrayNode();
        ObjectNode statement = mapper.createObjectNode();
        statement.put("Effect", "Allow");

        ArrayNode actions = mapper.createArrayNode();
        actions.add("logs:CreateLogGroup");
        actions.add("logs:CreateLogStream");
        actions.add("logs:PutLogEvents");
        statement.set("Action", actions);

        statement.put("Resource", "arn:aws:logs:" + getRegion()
                + ":*:log-group:/aws/lambda/" + getPrefix() + "*");

        statements.add(statement);
        policy.set("Statement", statements);

        return policy.toString();
    }

    private String createDynamoDBPolicy(final String tableArn) {
        ObjectNode policy = mapper.createObjectNode();
        policy.put("Version", "2012-10-17");

        ArrayNode statements = mapper.createArrayNode();
        ObjectNode statement = mapper.createObjectNode();
        statement.put("Effect", "Allow");

        ArrayNode actions = mapper.createArrayNode();
        actions.add("dynamodb:GetItem");
        actions.add("dynamodb:PutItem");
        actions.add("dynamodb:UpdateItem");
        actions.add("dynamodb:DeleteItem");
        actions.add("dynamodb:Query");
        actions.add("dynamodb:Scan");
        statement.set("Action", actions);

        ArrayNode resources = mapper.createArrayNode();
        resources.add(tableArn);
        resources.add(tableArn + "/index/*");
        statement.set("Resource", resources);

        statements.add(statement);
        policy.set("Statement", statements);

        return policy.toString();
    }

    private String createS3Policy(final String bucketArn) {
        ObjectNode policy = mapper.createObjectNode();
        policy.put("Version", "2012-10-17");

        ArrayNode statements = mapper.createArrayNode();
        ObjectNode statement = mapper.createObjectNode();
        statement.put("Effect", "Allow");

        ArrayNode actions = mapper.createArrayNode();
        actions.add("s3:GetObject");
        actions.add("s3:GetObjectVersion");
        statement.set("Action", actions);

        statement.put("Resource", bucketArn + "/*");

        statements.add(statement);
        policy.set("Statement", statements);

        return policy.toString();
    }

    public String getLambdaExecutionRoleArn() {
        return lambdaExecutionRole.getArn();
    }
}

File: archive/cdktf-ts/Pr5768/lib/iam-construct.ts
import { Construct } from 'constructs';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicy } from '@cdktf/provider-aws/lib/iam-role-policy';
import { EnvironmentConfig } from './environment-config';

export interface IamConstructProps {
  environmentSuffix: string;
  config: EnvironmentConfig;
  bucketArn: string;
  tableArn: string;
}

export class IamConstruct extends Construct {
  public readonly dataAccessRole: IamRole;
  public readonly dataAccessRoleArn: string;

  constructor(scope: Construct, id: string, props: IamConstructProps) {
    super(scope, id);

    const { environmentSuffix, config, bucketArn, tableArn } = props;

    // Create IAM role for data access with least-privilege
    this.dataAccessRole = new IamRole(this, 'DataAccessRole', {
      name: `data-access-role-${environmentSuffix}`,
      description: `Role for accessing ${config.environment} environment resources`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          },
        ],
      }),
      tags: {
        Name: `data-access-role-${environmentSuffix}`,
        Environment: config.environment,
        CostCenter: config.costCenter,
      },
    });

    this.dataAccessRoleArn = this.dataAccessRole.arn;

    // Attach least-privilege policy for S3 access
    new IamRolePolicy(this, 'S3AccessPolicy', {
      name: `s3-access-policy-${environmentSuffix}`,
      role: this.dataAccessRole.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              's3:GetObject',
              's3:PutObject',
              's3:DeleteObject',
              's3:ListBucket',
            ],
            Resource: [bucketArn, `${bucketArn}/*`],
          },
        ],
      }),
    });

    // Attach least-privilege policy for DynamoDB access
    new IamRolePolicy(this, 'DynamoDBAccessPolicy', {
      name: `dynamodb-access-policy-${environmentSuffix}`,
      role: this.dataAccessRole.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              'dynamodb:GetItem',
              'dynamodb:PutItem',
              'dynamodb:UpdateItem',
              'dynamodb:DeleteItem',
              'dynamodb:Query',
              'dynamodb:Scan',
            ],
            Resource: [tableArn, `${tableArn}/index/*`],
          },
        ],
      }),
    });

    // Add CloudWatch Logs permissions for Lambda
    new IamRolePolicy(this, 'CloudWatchLogsPolicy', {
      name: `cloudwatch-logs-policy-${environmentSuffix}`,
      role: this.dataAccessRole.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:PutLogEvents',
            ],
            Resource: 'arn:aws:logs:*:*:*',
          },
        ],
      }),
    });
  }
}
